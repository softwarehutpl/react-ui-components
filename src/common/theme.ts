const theme: { [key: string]: { [key: string]: { [key: string]: string } } } = {
  colors: {
    primary: {
      base: '#000000',
      light: '#ffffff',
      dark: '#a8b3bd',
    },
    secondary: {
      base: '#3690e3',
      light: '#75b0e6',
      dark: '#034d91',
    },
    error: {
      base: '#f44336',
      light: '#f56c62',
      dark: '#a1140a',
    },
    warning: {
      base: '#faa72d',
      light: '#ffc36b',
      dark: '#995b00',
    },
    info: {
      base: '#7979d4',
      light: '#9797d1',
      dark: '#4c4cc7',
    },
    success: {
      base: '#4caf50',
      light: '#88c28a',
      dark: '#124f14',
    },
    disabled: {
      base: '#d9d9d9',
      light: '#f0eded',
      dark: '#8c8989',
    },
  },
  // TODO: change it in dedicated task for that (change ease-out for example and add more options)
  transitions: {
    all: {
      fast: 'all 0.5s ease-out',
      mid: 'all 2s ease-out',
      slow: 'all 4s ease-out',
    },
    background: {
      fast: 'background-color 0.5s ease-out',
      mid: 'background-color 2s ease-out',
      slow: 'background-color 4s ease-out',
    },
    fontColor: {
      fast: 'color 0.5s ease-out',
      mid: 'color 2s ease-out',
      slow: 'color 4s ease-out',
    }
  }
};

export default theme;
