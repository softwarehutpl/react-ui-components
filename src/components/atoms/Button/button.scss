.test-class {
  color: yellowgreen;
  background-color: green;
  border-color: yellowgreen;
}
