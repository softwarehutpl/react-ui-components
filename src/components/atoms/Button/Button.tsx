import React from 'react';
import styled from 'styled-components';
import './button.scss';

type Props = {
  buttonTitle: string,
  onClick: () => void,
  color: string,
  className: string,
  disabled: boolean,
  fontColor?: string,
  noBorder: boolean,
  borderColor?: string,
  borderWidth: number,
  backgroundColor?: string,
}

const Button = (
  {
    buttonTitle,
    onClick,
    className,
    disabled,
  }: Props
) => (
  <button
    className={className}
    onClick={onClick}
    disabled={disabled}
  >
    { buttonTitle }
  </button>
);

const defaultProps = {
  color: 'primary',
  noBorder: false,
  disabled: false,
  className: '',
  borderWidth: 2,
};

Button.defaultProps = defaultProps;

export default styled(Button)`${({ 
  theme,
  color = defaultProps.color,
  backgroundColor,
  fontColor,
  noBorder = defaultProps.noBorder,
  borderColor,
  borderWidth = defaultProps.borderWidth,
}) => ({
  'background-color': backgroundColor || theme.colors[color].base,
  'color': fontColor || theme.colors[color].dark,
  'border': noBorder ? 'none' : 'solid',
  'border-color': borderColor || theme.colors[color].dark,
  'border-width': borderWidth + 'px',
  'border-radius': '10px',
  'outline': 'none',
  'padding': '15px',
  'text-transform': 'uppercase',
  'font-weight': 'bold',
  '&:hover': {
    'cursor': 'pointer',
    'background-color': theme.colors[color].light,
  },
  '&[disabled]': {
    'background-color': theme.colors.disabled.base,
    'border-color': theme.colors.disabled.dark,
    'color': '#a6a6a6',
    '&:hover': {
      'cursor': 'auto',
    }
  }
})}`
