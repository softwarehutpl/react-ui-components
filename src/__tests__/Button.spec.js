import React from 'react';
import { create, act } from 'react-test-renderer';
import 'jest-styled-components';
import Button from '../components/atoms/Button/Button';
import theme from '../common/theme';

describe('Button component', () => {
  it('should render correctly', () => {
    const tree = create(<Button theme={theme} buttonTitle="TITLE" onClick={() => {}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render button component with proper buttonTitle', () => {
    const component = create(<Button buttonTitle="TITLE" theme={theme} />);
    const instance = component.root;
    const button = instance.findByType('button');
    expect(button.props.children).toBe('TITLE');
  });

  it('should call button onClick method', () => {
    const mockOnClick = jest.fn();
    const component = create(<Button onClick={mockOnClick} theme={theme} />);
    const instance = component.root;
    const button = instance.findByType(Button);
    act(() => {
      button.props.onClick();
    });
    expect(mockOnClick.mock.calls.length).toEqual(1);
  });

  it('should render button with custom class', () => {
    const component = create(<Button theme={theme} className="test-class" />);
    const button = component.toJSON();
    expect(button.props.className).toEqual(expect.stringContaining('test-class'));
  });

  it('should render button component with selected theme styling', () => {
    const component = create(<Button theme={theme} color="secondary" />);
    const button = component.toJSON();
    expect(button).toHaveStyleRule('background-color', '#3690e3');
    expect(button).toHaveStyleRule('color', '#75b0e6');
    expect(button).toHaveStyleRule('border-color', '#75b0e6');
    expect(button).toHaveStyleRule('background-color', '#034d91', { modifier: ':hover' });
  });

  it('should render disabled button', () => {
    const component = create(<Button theme={theme} disabled />);
    const button = component.toJSON();
    expect(button.props.disabled).toBe(true);
  });

  it('should render button component with color based on fontColor prop', () => {
    const component = create(<Button theme={theme} fontColor="red" />);
    const button = component.toJSON();
    expect(button).toHaveStyleRule('color', 'red');
  });

  it('should render button with custom hover styles', () => {
    const component = create(<Button theme={theme} hoverBackgroundColor="yellow" hoverBorderColor="red" />);
    const button = component.toJSON();
    expect(button).toHaveStyleRule('border-color', 'red', { modifier: ':hover' });
    expect(button).toHaveStyleRule('background-color', 'yellow', { modifier: ':hover' });
  });

  it('should render button with custom border', () => {
    const component = create(
      <Button
        theme={theme}
        noBorder={false}
        borderColor="blue"
        borderWidth={3}
        borderRadius={10}
      />
    );
    const button = component.toJSON();
    expect(button).toHaveStyleRule('border', 'solid');
    expect(button).toHaveStyleRule('border-color', 'blue');
    expect(button).toHaveStyleRule('border-width', '3px');
    expect(button).toHaveStyleRule('border-radius', '10px');
  });

  it('should render button with default border', () => {
    const component = create(<Button theme={theme} noBorder={false} />);
    const button = component.toJSON();
    expect(button).toHaveStyleRule('border', 'solid');
    expect(button).toHaveStyleRule('border-color', '#ffffff');
    expect(button).toHaveStyleRule('border-width', '2px');
    expect(button).toHaveStyleRule('border-radius', '4px');
  });

  it('should render button component with custom padding and margin', () => {
    const component = create(<Button theme={theme} padding={30} margin={10} />);
    const button = component.toJSON();
    expect(button).toHaveStyleRule('padding', '30px');
    expect(button).toHaveStyleRule('margin', '10px');
  });
});
