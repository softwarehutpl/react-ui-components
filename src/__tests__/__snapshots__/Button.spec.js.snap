// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render correctly 1`] = `
.c0 {
  background-color: #000000;
  color: #ffffff;
  border: none;
  border-color: #ffffff;
  border-width: 2px;
  border-radius: 4px;
  margin: 0px;
  padding: 15px;
  text-transform: uppercase;
  font-weight: bold;
  outline: none;
}

.c0:hover {
  cursor: pointer;
  background-color: #a8b3bd;
}

.c0[disabled] {
  background-color: #d9d9d9;
  border-color: #8c8989;
  color: #a6a6a6;
}

.c0[disabled]:hover {
  cursor: auto;
}

<button
  className="c0"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  TITLE
</button>
`;
